File: Class3/절댓값힙.cpp
#include<iostream>
#include<queue>
#include<functional>
using namespace std;

int main() {
    cin.tie(0);
    ios::sync_with_stdio(0);

    int N;
    cin >> N;

    auto cmp = [](pair<int, int> left, pair<int, int> right) {
        if (left.first == right.first) {
            return left.second > right.second; // Push negative integers first if absolute values are the same
        }
        return left.first > right.first;
    };

    priority_queue<pair<int, int>, vector<pair<int, int>>, decltype(cmp)> absHeap(cmp);

    while(N--) {
        int x;
        cin >> x;
        if (x == 0) {
            if (absHeap.empty()) {
                cout << "result : " << 0 << '\n';
            } else {
                cout << "result : " << absHeap.top().second << '\n';
                absHeap.pop();
            }
        } else {
            absHeap.push({abs(x), x}); // Push pair of absolute value and original value
        }
    }

    return 0;
}

File: Data Structure/Queue/절댓값 힙.py
import sys
from queue import PriorityQueue

N = int(sys.stdin.readline())
pq = PriorityQueue()

for _ in range(N):
    num = int(sys.stdin.readline())
    if num == 0:
        if pq.empty():
            print(0)
        else:
            print(pq.get()[1])
    else:
        pq.put((abs(num), num))